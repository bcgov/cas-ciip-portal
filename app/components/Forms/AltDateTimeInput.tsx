/**
 * AltDateTimeWidget modified from rjsf-team/react-jsonschema-form licensed under Apache 2.0:
 *
 * https://github.com/rjsf-team/react-jsonschema-form/blob/v1.8.1/src/components/widgets/AltDateTimeWidget.js
 * https://github.com/rjsf-team/react-jsonschema-form/blob/v1.8.1/LICENSE.md
 *
 * Changelog:
 * - Name changed to AltDateTimeInput
 * - propTypes removed
 * - linter formatting
 */

import React from 'react';
import AltDateInput from './AltDateInput';

const AltDateTimeInput = (props) => {
  const {AltDateInput} = props.registry.widgets;
  return <AltDateInput time {...props} />;
};

AltDateTimeInput.defaultProps = {
  ...AltDateInput.defaultProps,
  time: true
};

export default AltDateTimeInput;
