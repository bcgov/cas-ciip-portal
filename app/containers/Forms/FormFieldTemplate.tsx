import React from 'react';
import {Col, Form} from 'react-bootstrap';
import {FieldTemplateProps} from 'react-jsonschema-form';

const FormFieldTemplate: React.FunctionComponent<FieldTemplateProps> = ({
  id,
  label,
  help,
  required,
  description,
  errors,
  children,
  schema,
  classNames,
  uiSchema,
  hidden
}) => {
  if (hidden) return <div className="hidden">{children}</div>;

  const helpText = uiSchema['ui:helplink'] ? (
    <a
      className="font-italic"
      href={uiSchema['ui:helplink']}
      target="_blank"
      rel="noopener noreferrer"
    >
      {help}
    </a>
  ) : (
    <span className="font-italic text-muted">{help}</span>
  );

  if (schema.type === 'array')
    return (
      <Col xs={12}>
        {description}
        {helpText}
        {errors}
        {children}
      </Col>
    );
  if (schema.type === 'object')
    return (
      <Col xs={12}>
        <Form.Row>
          {description}
          {errors}
          {helpText}
          {children}
        </Form.Row>
      </Col>
    );

  const xs = uiSchema['ui:col-xs'] || 12;
  const sm = uiSchema['ui:col-sm'] || xs;
  const md = uiSchema['ui:col-md'] || (sm < 6 ? sm : 6);
  const lg = uiSchema['ui:col-lg'] || md;
  const xl = uiSchema['ui:col-xl'] || lg;

  return (
    <Form.Group
      as={Col}
      xs={xs}
      sm={sm}
      md={md}
      lg={lg}
      xl={xl}
      className={classNames}
    >
      <Form.Label htmlFor={id}>
        {label}
        {required ? <>&nbsp;*</> : null}
      </Form.Label>
      {description}
      {children}
      {errors}
      {helpText}
    </Form.Group>
  );
};

export default FormFieldTemplate;
