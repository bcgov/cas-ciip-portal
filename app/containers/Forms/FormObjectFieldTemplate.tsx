import React from 'react';
import {Col} from 'react-bootstrap';

const FormObjectFieldTemplate = ({
  TitleField,
  properties,
  title,
  description
}) => {
  return (
    <>
      {title && (
        <Col xs={12}>
          <TitleField title={title} />
        </Col>
      )}
      {properties.map(prop => (
        <React.Fragment key={prop.content.key}>{prop.content}</React.Fragment>
      ))}
      {description}
    </>
  );
};

export default FormObjectFieldTemplate;
