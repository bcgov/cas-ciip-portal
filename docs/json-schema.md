# Usage of JSON schemas in the CIIP portal

This application's source code relies extensively on the [JSON Schema specification] to define the shape and validation of the data that is part of a CIIP application.

## Rendering of the data using @rjsf/core

In order to display the data to the user, either when displaying the forms where industry reporter input their data, or when rendering the contents of a CIIP application in a table, we rely on the [@rjsf/core] project.

TODO: explain format of objects ({schema, uiSchema, customFormats, customErrors})

### TODO: support of formulas in forms

[json schema specification]: https://json-schema.org/
[@rjsf/core]: https://github.com/rjsf-team/@rjsf/core
