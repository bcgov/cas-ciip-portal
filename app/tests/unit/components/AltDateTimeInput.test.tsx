import React from 'react';
import {shallow} from 'enzyme';

import JsonSchemaForm from 'react-jsonschema-form';
import {JSONSchema6} from 'json-schema';
import AltDateTimeInput from 'components/Forms/AltDateTimeInput';
import FormObjectFieldTemplate from 'containers/Forms/FormObjectFieldTemplate';
import FormFieldTemplate from 'containers/Forms/FormFieldTemplate';

describe('AltDateInput', () => {
  it('Should match last accepted snapshot: AltDateTime (both date and time)', async () => {
    const schema = {
      applicationOpenTime: {
        title: 'Application Open Time',
        type: 'string'
      }
    };
    const uiSchema = {
      applicationOpenTime: {
        'ui:widget': 'AltDateTimeInput',
        'ui:options': {
          hideNowButton: true,
          hideClearButton: true
        }
      }
    };
    const r = shallow(
      <JsonSchemaForm
        omitExtraData
        liveOmit
        noHtml5Validate
        schema={schema as JSONSchema6}
        uiSchema={uiSchema}
        formData={{}}
        FieldTemplate={FormFieldTemplate}
        ObjectFieldTemplate={FormObjectFieldTemplate}
        widgets={{AltDateTimeInput}}
        showErrorList={false}
      />
    );
    expect(r).toMatchSnapshot();
  });
});
