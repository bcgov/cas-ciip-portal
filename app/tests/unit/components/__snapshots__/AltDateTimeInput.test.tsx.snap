// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AltDateInput Should match last accepted snapshot: AltDateTime (both date and time) 1`] = `
<form
  className="rjsf"
  noValidate={true}
  onSubmit={[Function]}
>
  <SchemaField
    autofocus={false}
    disabled={false}
    errorSchema={Object {}}
    formData={Object {}}
    idSchema={
      Object {
        "$id": "root",
      }
    }
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    readonly={false}
    registry={
      Object {
        "ArrayFieldTemplate": undefined,
        "FieldTemplate": [Function],
        "ObjectFieldTemplate": [Function],
        "definitions": Object {},
        "fields": Object {
          "AnyOfField": [Function],
          "ArrayField": [Function],
          "BooleanField": [Function],
          "DescriptionField": [Function],
          "NullField": [Function],
          "NumberField": [Function],
          "ObjectField": [Function],
          "OneOfField": [Function],
          "SchemaField": [Function],
          "StringField": [Function],
          "TitleField": [Function],
          "UnsupportedField": [Function],
        },
        "formContext": Object {},
        "widgets": Object {
          "AltDateTimeInput": [Function],
          "AltDateTimeWidget": [Function],
          "AltDateWidget": [Function],
          "BaseInput": [Function],
          "CheckboxWidget": [Function],
          "CheckboxesWidget": [Function],
          "ColorWidget": [Function],
          "DateTimeWidget": [Function],
          "DateWidget": [Function],
          "EmailWidget": [Function],
          "FileWidget": [Function],
          "HiddenWidget": [Function],
          "PasswordWidget": [Function],
          "RadioWidget": [Function],
          "RangeWidget": [Function],
          "SelectWidget": [Function],
          "TextWidget": [Function],
          "TextareaWidget": [Function],
          "URLWidget": [Function],
          "UpDownWidget": [Function],
        },
      }
    }
    safeRenderCompletion={false}
    schema={
      Object {
        "applicationOpenTime": Object {
          "title": "Application Open Time",
          "type": "string",
        },
      }
    }
    uiSchema={
      Object {
        "applicationOpenTime": Object {
          "ui:options": Object {
            "hideClearButton": true,
            "hideNowButton": true,
          },
          "ui:widget": "AltDateTimeInput",
        },
      }
    }
  />
  <div>
    <button
      className="btn btn-info"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
